### DISCUSSION TITLE
Dynamic `cmap`

### DISCUSSION THREAD

**Discussion by stla** [pyvista/pyvista#2078] [Show and tell]
"Show and tell" category :-) I like it:

![](https://github.com/stla/PyVistaMiscellanous/raw/main/HopfTorusMovingFlag.gif)

The "trick":

```python
dists = np.linalg.norm(mesh.points, axis=1)
dists = 2*pi * (dists - dists.min()) / (dists.max() - dists.min())


t_ = np.linspace(0, 2*pi, 60, endpoint=False)
for i, t in enumerate(t_):
    pngname = "zzpic%03d.png" % i
    pltr = pv.Plotter(window_size = [512, 512], off_screen=True)
    pltr.set_focus([0, 0, 0])
    pltr.set_position((75, 0, 0))
    mesh["dist"] = np.sin(dists - t) # <---------------------------- the trick
    pltr.add_mesh(
        mesh, smooth_shading=True, cmap="flag", show_scalar_bar=False,
        specular=10
    )
    pltr.show(screenshot=pngname)
```

[Full code](https://github.com/stla/PyVistaMiscellanous/blob/main/HopfTorus_MovingFlag.py).

**Comment by banesullivan** [comment]
Super cool!! Thanks for sharing, @stla! If you have the time, it'd be great to perhaps add this to the examples gallery

**Comment by stla** [comment]
Will take a look at your repo to see how you handle the examples.

With `cmap="nipy_spectral"`, cool as well:

![HopfTorusMovingNipySpectral](https://user-images.githubusercontent.com/4466543/151410360-0c4aabf8-6403-4e7d-bbd9-94c032e865b1.gif)

**Reply by banesullivan** [reply]
Take a look at the `examples/` directory and you can copy the structure of the `.py` files in there - we use sphinx-gallery to generate them. FYI, there is current work to improve docs building so don't be discouraged if you can't build the documentation locally. Things should be streamlined after #1910

**Comment by stla** [comment]
Ok, but how do you generate some GIF files?

Just written a short [blog post](https://laustep.github.io/stlahblog/posts/PyVistaDynamicCmap.html) about this dynamic texture.

**Reply by tkoyama010** [reply]
This example maybe help you to add [animation](https://dev.pyvista.org/examples/01-filter/flying_edges.html#animate-barth-sextic). Feel free to PRing.

**Comment by stla** [comment]
Nice! I always did my animations "manually", saving each frame and mounting them with ImageMagick or gifski. I didn't know this technique.

```python
from math import pi
import numpy as np
import pyvista as pv


def Gamma(t):
    alpha = pi / 2 - (pi / 2 - 0.44) * np.cos(3 * t)
    beta = t + 0.44 * np.sin(6 * t)
    return np.array(
        [
          np.sin(alpha) * np.cos(beta), 
          np.sin(alpha) * np.sin(beta), 
          np.cos(alpha)
        ]
    )


def HopfFiber(p, phi):
    return (
        np.array(
            [
                (1 + p[2]) * np.cos(phi),
                p[0] * np.sin(phi) - p[1] * np.cos(phi),
                p[0] * np.cos(phi) + p[1] * np.sin(phi),
                (1 + p[2]) * np.sin(phi),
            ]
        )
        / np.sqrt(2 * (1 + p[2]))
    )


def Stereo(q):
    return q[0:3] / (1 - q[3])


def F(t, phi):
    return Stereo(HopfFiber(Gamma(t), phi))


def HTmesh(nu=400, nv=200):
    angle_u = np.linspace(-pi, pi, nu)
    angle_v = np.linspace(0, pi, nv)
    u, v = np.meshgrid(angle_u, angle_v)
    x, y, z = F(u, v)
    grid = pv.StructuredGrid(x, y, z)
    return grid.extract_geometry().clean(tolerance=1e-6)


mesh = HTmesh()
dists = np.linalg.norm(mesh.points, axis=1)
dists = 2 * pi * (dists - dists.min()) / (dists.max() - dists.min())

pltr = pv.Plotter(window_size=[512,512])
pltr.set_focus([0, 0, 0])
pltr.set_position((40, 0, 0))
pltr.add_mesh(
    mesh,
    scalars=np.sin(dists),
    smooth_shading=True,
    specular=10,
    cmap="flag",
    show_scalar_bar=False,
)
pltr.open_gif("Hopf_torus.gif")

for t in np.linspace(0, 2 * pi, 60, endpoint=False):
    pltr.update_scalars(np.sin(dists - t), render=False)
    pltr.write_frame()

pltr.show()
```

**Comment by stla** [comment]
![](https://github.com/stla/PyVistaMiscellanous/raw/main/twentyHopfTori3.gif)

**Reply by RichardScottOZ** [reply]
Now  I want to make a Beholder like that.
